### PR TITLE
Improve scan operation

### PR DESCRIPTION
**Pull request by okuta**
This PR reduces CUDA kernel launch count.

### CODE REVIEW

**Comment by emcastillo** [comment]
Thanks!

File: cupy/core/_routines_indexing.pyx [hunk 1]
```diff
@@ -63,14 +63,18 @@ cdef tuple _ndarray_nonzero(ndarray self):
...
-        nonzero = cupy.core.not_equal(r, 0, ndarray(r.shape, dtype))
-        del r
-        scan_index = _math.scan(nonzero, op=_math.scan_op.SCAN_SUM)
+        if self.dtype == numpy.bool_:
+            nonzero = self.ravel()
+        else:
+            nonzero = cupy.core.not_equal(self, 0)
+            nonzero = nonzero.ravel()
+        scan_index = _math.scan(nonzero, op=_math.scan_op.SCAN_SUM,
+                                dtype=dtype)
```
Comment (leofang) [inline cupy/core/_routines_indexing.pyx:77]: Do we always have to use `int64` as the dtype here?

Comment (okuta) [reply cupy/core/_routines_indexing.pyx]: Thanks, I fixed.

**Comment by okuta** [comment]
jenkins, test this please.

File: cupy/core/_routines_indexing.pyx (hunk 1 above)
Comment (emcastillo) [reply cupy/core/_routines_indexing.pyx]: I think @leofang meant that we could use int32 for arrays with less elements than 2**31-1

Comment (okuta) [reply cupy/core/_routines_indexing.pyx]: Because numpy returns `numpy.int64` index array.

Comment (leofang) [reply cupy/core/_routines_indexing.pyx]: I see. But, it seems scan is just an intermediate step, and the actual returned array is computed in `_nonzero_kernel*`? Would it be possible to defer the generation of an int64 array to there?

Comment (leofang) [reply cupy/core/_routines_indexing.pyx]: (Not saying it'd be faster! Just curious...Feel free to ignore me 😅)

Comment (okuta) [reply cupy/core/_routines_indexing.pyx]: Ah, I understand this function only returns `dst`

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by leofang** [comment]
Cool!

**Pull request merged**